File: Tools/BuildTool/Source/Toolchains/ToolchainMSVC.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Runtime.InteropServices;
using System.Text;

enum EMSVCArchetecture 
{
    x64,
    x86
};

class MSVCBuildPlatform : BuildPlatform 
{
    public EMSVCArchetecture Architecture;
}

public class ToolchainMSVC : IToolchainInterface<ToolchainMSVC>
{
    public override List<BuildPlatform> GetPlatforms()
    {
        return new List<BuildPlatform>()
        {
            new MSVCBuildPlatform() { Architecture = EMSVCArchetecture.x64, platformName = "Windows x64", preprocessorDefines = { "FARO_OS_WINDOWS", "FARO_ARCH_X64" }, linkerLibraries = {} },
            new MSVCBuildPlatform() { Architecture = EMSVCArchetecture.x86, platformName = "Windows x86", preprocessorDefines = { "FARO_OS_WINDOWS", "FARO_ARCH_X86" }, linkerLibraries = {} }
        };
    }

    private string objDir = "";
    private string libPath = "";

    private string msvcRoot = "";
    private string msvcTools = "";

    private string windowsSdkInclude = "";
    private string windowsSdkLib = "";

    public override bool PrepareModuleForBuild(ModuleManifest manifest, BuildPlatform target)
    {
        MSVCBuildPlatform buildPlatform = (MSVCBuildPlatform)target;

        int windowsKitCount = Utility.CountWindowsKits();
        if (windowsKitCount > 0)
        {
            string root = Marshal.PtrToStringAnsi(Utility.GetWindowsKitRoot(0));
            string version = Marshal.PtrToStringAnsi(Utility.GetWindowsKitVersion(0));

            windowsSdkInclude = root + "\\Include\\" + version;
            windowsSdkLib = root + "\\Lib\\" + version;
        }

        int MSVCCount = Utility.CountMSVC();
        if (MSVCCount > 0)
        {
            msvcRoot = Marshal.PtrToStringAnsi(Utility.GetMSVCRoot(0));
        }

        if (windowsSdkLib == "" || windowsSdkInclude == "" || !Directory.Exists(windowsSdkLib) || !Directory.Exists(windowsSdkInclude))
        {
            Utility.PrintLine("Invalid WindowsKit directory");
            return false;
        }

        if (msvcRoot == "" || !Directory.Exists(msvcRoot))
        {
            Utility.PrintLine("Invalid MSVC directory");
            return false;
        }

        msvcTools = msvcRoot + "\\bin\\";
        switch (buildPlatform.Architecture)
        {
            case EMSVCArchetecture.x64:
                {
                    msvcTools += "Hostx64\\x64";
                    break;
                }
            case EMSVCArchetecture.x86:
                {
                    msvcTools += "Hostx86\\x86";
                    break;
                }
        }

        if (!Directory.Exists(windowsSdkInclude) || !Directory.Exists(windowsSdkLib) || !Directory.Exists(msvcTools))
        {
            Utility.PrintLine("Invalid WindowsKit");
            return false;
        }

        objDir = manifest.buildRoot + "\\Obj\\";
        Directory.CreateDirectory(objDir);

        libPath = GetModuleLibrary(manifest);
        Directory.CreateDirectory(GetLibDirForModule(manifest));

        return true;
    }

    private string GetObjForSource(string file)
    {
        return objDir + "\\" + Path.GetFileNameWithoutExtension(file) + ".obj";
    }

    public override bool BuildSource(string sourceFile, List<string> includePaths, List<string> preprocessorDefines)
    {
        string includes = "";

        includes += " /I\"" + msvcRoot + "\\include\"";
        includes += " /I\"" + windowsSdkInclude + "\\shared\"";
        includes += " /I\"" + windowsSdkInclude + "\\ucrt\"";
        includes += " /I\"" + windowsSdkInclude + "\\winrt\"";
        includes += " /I\"" + windowsSdkInclude + "\\um\"";

        foreach (string include in includePaths) 
        {
            includes += " /I\"" + include + "\"";
        }

        string defines = "";
        foreach (string define in preprocessorDefines)
        {
            includes += " /D" + define;
        }

        string clExe = "\"" + msvcTools + "\\cl.exe\"";
        string msvcDrive = msvcRoot.Substring(0, 1);
        string outputFile = GetObjForSource(sourceFile);

        string log = "";
        int result = ExecuteCommand(msvcDrive + ": & " + GetEnvCommand() + " & " + clExe + " /c /FC /nologo /EHsc /Z7 /Od " + defines + " " + sourceFile + " /Fo\"" + outputFile + "\" " + includes, out log);
        
        //Format, trim and print output message
        if (log.Length > 0)
        {
            log = log.Replace("\r", "");

            string header = Path.GetFileName(sourceFile) + "\n";
            if (log.StartsWith(header))
            {
                log = log.Substring(header.Length);
            }

            if (log.Length > 0)
            {
                Utility.PrintLine(log);
            }
        }

        return result == 0;
    }

    public override bool LinkLibrary(ModuleManifest module)
    {
        string objs = "";
        foreach (string sourceFile in module.sourceFiles)
        {
            objs += " \"" + GetObjForSource(sourceFile) + "\"";
        }

        string libExe = "\"" + msvcTools + "\\lib.exe\"";
        string msvcDrive = msvcRoot.Substring(0, 1);

        string log = "";
        int result = ExecuteCommand(msvcDrive + ": & " + GetEnvCommand() + " & " + libExe + " /nologo /OUT:\"" + libPath + "\" " + objs, out log);

        //Format, trim and print output message
        if (log.Length > 0)
        {
            log = log.Replace("\r", "");

            if (log.Length > 0)
            {
                Utility.PrintLine(log);
            }
        }

        return result == 0;
    }

    public override bool LinkExecutable(ModuleManifest module)
    {
        string libs = "";
        string libDirectories = "";
        string moduleLibs = "";

        libs += " \"user32.lib\"";

        //TODO switch debug and release
        libs += " \"libucrtd.lib\"";
        libs += " \"libvcruntimed.lib\"";

        //TODO Set correct architecture
        libDirectories += " /LIBPATH:\"" + windowsSdkLib + "\\ucrt\\x64\"";
        libDirectories += " /LIBPATH:\"" + windowsSdkLib + "\\um\\x64\"";
        libDirectories += " /LIBPATH:\"" + msvcRoot + "\\lib\\x64\"";

        //TODO link against other modules (/WHOLEARCHIVE)
        //TODO link against dependencies

        string linkExe = "\"" + msvcTools + "\\link.exe\"";
        string msvcDrive = msvcRoot.Substring(0, 1);

        string outDir = module.buildRoot + "\\bin";
        Directory.CreateDirectory(outDir);
        string outputFile = outDir + "\\" + module.name + ".exe";

        string log = "";
        int result = ExecuteCommand(msvcDrive + ": & " + GetEnvCommand() + " & " + linkExe + " /NOLOGO /DEBUG /SUBSYSTEM:WINDOWS /MACHINE:X64 /OUT:\"" + outputFile + "\" " + libs + libDirectories + moduleLibs, out log);

        //Format, trim and print output message
        if (log.Length > 0)
        {
            log = log.Replace("\r", "");

            if (log.Length > 0)
            {
                Utility.PrintLine(log);
            }
        }
        return result == 0;
    }

    private string GetEnvCommand()
    {
        //TODO add MSVC IDE path
        return "set \"path=C:\\Windows\\System32;\"";
    }

    private string GetLibDirForModule(ModuleManifest manifest)
    {
        return manifest.buildRoot + "\\lib";
    }

    private string GetModuleLibrary(ModuleManifest manifest)
    {
        return GetLibDirForModule(manifest) + "\\Windows.lib";
    }
}

File: Tools/BuildTool/Source/Toolchains/ToolchainMSVC.cpp
#include "ToolchainMSVC.hpp"
#include <Utility.hpp>
#include "MSVCInfo.hpp"
#include "WindowsKitInfo.hpp"

std::vector<BuildPlatform*> ToolchainMSVC::GetPlatforms()
{
    if (platforms.empty())
    {
        platforms.push_back(new MSVCBuildPlatform{ "WindowsX64", { "FARO_OS_WINDOWS", "FARO_ARCH_X64" }, {}, EMSVCArchitecture::x64 });
        platforms.push_back(new MSVCBuildPlatform{ "WindowsX86", { "FARO_OS_WINDOWS", "FARO_ARCH_X86" }, {}, EMSVCArchitecture::x86 });
    }

    return platforms;
}

bool ToolchainMSVC::PrepareModuleForBuild(ModuleManifest& manifest, BuildPlatform* target, BuildType configuration)
{
    MSVCBuildPlatform* buildPlatform = reinterpret_cast<MSVCBuildPlatform*>(target);

    const std::vector<WindowsKit>& windowsKits = GetWindowsKits();
    if (!windowsKits.empty())
    {
        std::string root = windowsKits[0].Root;
        std::string version = windowsKits[0].Version;

        windowsSdkInclude = root + "\\Include\\" + version;
        windowsSdkLib = root + "\\Lib\\" + version;
    }

    const std::vector<MSVCVersion>& MSVCInstallations = GetMSVCInstallations();
    if (!MSVCInstallations.empty())
    {
        msvcRoot = MSVCInstallations[0].Root;
    }

    if (windowsSdkLib == "" || windowsSdkInclude == "" || !std::filesystem::exists(windowsSdkLib) || !std::filesystem::exists(windowsSdkInclude))
    {
        Utility::PrintLine("Invalid WindowsKit directory");
        return false;
    }

    if (msvcRoot == "" || !std::filesystem::exists(msvcRoot))
    {
        Utility::PrintLine("Invalid MSVC directory");
        return false;
    }

    msvcTools = msvcRoot / "bin";
    switch (buildPlatform->Architecture)
    {
    case EMSVCArchitecture::x64:
        {
            msvcTools /= "Hostx64\\x64";
            break;
        }
    case EMSVCArchitecture::x86:
        {
            msvcTools /= "Hostx86\\x86";
            break;
        }
    }

    if (!std::filesystem::exists(windowsSdkInclude) || !std::filesystem::exists(windowsSdkLib) || !std::filesystem::exists(msvcTools))
    {
        Utility::PrintLine("Invalid WindowsKit");
        return false;
    }

    objDir = GetObjDirectory(manifest, target, configuration);
    Utility::EnsureDirectory(objDir);

    return true;
}

bool ToolchainMSVC::BuildSource(ModuleManifest& manifest, BuildPlatform* target, BuildType configuration, std::filesystem::path sourceFile, std::vector<std::filesystem::path> includePaths, std::vector<std::string> preprocessorDefines)
{
    std::string includes = "";

    includes += " /I\"" + msvcRoot.string() + "\\include\""; //TODO Don't define here but expose to ensure they're in-line with the generated project
    includes += " /I\"" + windowsSdkInclude.string() + "\\shared\"";
    includes += " /I\"" + windowsSdkInclude.string() + "\\ucrt\"";
    includes += " /I\"" + windowsSdkInclude.string() + "\\winrt\"";
    includes += " /I\"" + windowsSdkInclude.string() + "\\um\"";

    for (std::filesystem::path& include : includePaths)
    {
        includes += " /I\"" + include.string() + "\"";
    }

    std::string defines;
    for (std::string& define : preprocessorDefines)
    {
        defines += " /D" + define;
    }

    std::string compilerFlags;
    switch (configuration)
    {
    case Debug:
    case Development: { compilerFlags = " /DDEBUG /D_DEBUG /D_MT /D_CRTDBG_MAP_ALLOC /MTd /Od /Zi"; break; }
    case Release: { compilerFlags = " /D_MT /D_CRTDBG_MAP_ALLOC /MT /O2"; break; }

    case ENUMSIZE:;
    default:;
    }

    std::filesystem::path clExe = msvcTools.string() + "\\cl.exe";
    std::filesystem::path msvcDrive = msvcRoot.string().substr(0, 1);
    std::filesystem::path outputFile = GetObjPath(manifest, target, configuration, sourceFile);
    Utility::EnsureDirectory(outputFile.parent_path());

    std::string log = "";
    int result = ExecuteCommand(msvcDrive.string() + ": & " + GetEnvCommand() + " & \"" + clExe.string() + "\" /c /FC /nologo /EHsc " + defines + " " + compilerFlags + " " + sourceFile.string() + " /Fo\"" + outputFile.string() + "\" " + includes, log);
        
    // Format, trim and print output message
    if (!log.empty())
    {
        log.erase(std::remove(log.begin(), log.end(), '\r'), log.end());

        std::string header = sourceFile.filename().string() + "\n";
        if (log.substr(0, header.size()) == header)
        {
            log = log.substr(header.size());
        }

        if (!log.empty())
        {
            Utility::PrintLine(log);
        }
    }

    return result == 0;
}

bool ToolchainMSVC::LinkLibrary(ModuleManifest& manifest, BuildPlatform* target, BuildType configuration, std::vector<std::filesystem::path> sourceFiles)
{
    std::string objs = "";
    for (std::filesystem::path& sourceFile : sourceFiles)
    {
        objs += " \"" + GetObjPath(manifest, target, configuration, sourceFile).string() + "\"";
    }

    std::filesystem::path libExe = msvcTools / "lib.exe";
    std::filesystem::path msvcDrive = msvcRoot.string().substr(0, 1);

    std::filesystem::path libPath = GetLibPath(manifest, target, configuration);
    Utility::EnsureDirectory(libPath.parent_path());
    Utility::PrintLineD(libPath.string());

    std::string log = "";
    int result = ExecuteCommand(msvcDrive.string() + ": & " + GetEnvCommand() + " & \"" + libExe.string() + "\" /nologo /OUT:\"" + libPath.string() + "\" " + objs, log);

    // Format, trim and print output message
    if (!log.empty())
    {
        log.erase(std::remove(log.begin(), log.end(), '\r'), log.end());

        if (!log.empty())
        {
            Utility::PrintLine(log);
        }
    }

    return result == 0;
}

bool ToolchainMSVC::LinkExecutable(ModuleManifest& manifest, BuildPlatform* target, BuildType configuration, std::vector<std::filesystem::path> sourceFiles)
{
    // Get project object files
    std::string objs = "";
    for (std::filesystem::path& sourceFile : sourceFiles)
    {
        objs += " \"" + GetObjPath(manifest, target, configuration, sourceFile).string() + "\"";
    }

    std::string libs = "";
    std::string libDirectories = "";
    std::string moduleLibs = "";

    libs += " \"user32.lib\"";

    switch (configuration)
    {
    case Debug:
    case Development:
        {
            libs += " \"libucrtd.lib\"";
            libs += " \"libvcruntimed.lib\"";
            break;
        }
    case Release: 
        {
            libs += " \"libucrt.lib\"";
            libs += " \"libvcruntime.lib\"";
            break;
        }

    case ENUMSIZE:;
    default:;
    }

    //TODO Set correct architecture
    libDirectories += " /LIBPATH:\"" + windowsSdkLib.string() + "\\ucrt\\x64\"";
    libDirectories += " /LIBPATH:\"" + windowsSdkLib.string() + "\\um\\x64\"";
    libDirectories += " /LIBPATH:\"" + msvcRoot.string() + "\\lib\\x64\"";

    // Get dependency libraries
    for (ModuleManifest* module : GetAllModuleDependencies(manifest))
    {
        if (!module->IsCompatible(target)) continue;

        std::filesystem::path lib = GetLibPath(*module, target, configuration);
        Utility::PrintLineD("\t" + lib.string());
        moduleLibs += " /WHOLEARCHIVE:\""+ lib.string() + "\"";

        for (std::string& linkerLibrary : module->linkingLibraries)
        {
            libs += " \"" + linkerLibrary + "\"";
        }
    }

    std::filesystem::path linkExe = msvcTools / "link.exe";
    std::filesystem::path msvcDrive = msvcRoot.string().substr(0, 1);

    std::filesystem::path outputFile = GetExePath(manifest, target, configuration);
    Utility::EnsureDirectory(outputFile.parent_path());
    Utility::PrintLineD(outputFile.string());

    std::string flags = "/NOLOGO /SUBSYSTEM:CONSOLE /WX /MACHINE:X64";

    if (configuration == Debug || configuration == Development)
    {
        flags += " /DEBUG";
    }

    std::string log = "";
    int result = ExecuteCommand(msvcDrive.string() + ": & " + GetEnvCommand() + " & \"" + linkExe.string() + "\" " + flags + " /OUT:\"" + outputFile.string() + "\" " + objs + libs + libDirectories + moduleLibs, log);

    // Format, trim and print output message
    if (!log.empty())
    {
        log.erase(std::remove(log.begin(), log.end(), '\r'), log.end());

        if (!log.empty())
        {
            Utility::PrintLine(log);
        }
    }
    return result == 0;
}

std::string ToolchainMSVC::GetObjExtension()
{
    return "obj";
}

std::string ToolchainMSVC::GetLibExtension()
{
    return "lib";
}

std::string ToolchainMSVC::GetExeExtension()
{
    return "exe";
}

std::vector<std::filesystem::path> ToolchainMSVC::GetToolchainIncludes(BuildPlatform*, BuildType)
{
    return { "$(VC_IncludePath)", "$(WindowsSDK_IncludePath)" }; //TODO Replace by specific paths
}

std::string ToolchainMSVC::GetEnvCommand()
{
    //TODO Add MSVC IDE path
    return "set \"path=C:\\Windows\\System32;\"";
}
